docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,292 @@
+specific security context constraint setting or to generate settings using the following approach:
+
+**Pod with referenced service account**
+
+1.  Ensure that the user has permissions to reference the service account.
+2.  Retrieve the service account's security context constraints.
+3.  Ask the SecurityContextConstraintsAllocator to create the security context.  The allocator
+will not overwrite fields already set in the container.
+4.  Validate that the generated SecurityContext falls within the boundaries of the security context
+constraints and accept or reject the pod.
+
+**Pods with no service account**
+
+1.  Create a union of the user level permissions in the constraints.  This is the highest level
+under which they can create a pod if they are not referencing a service account.
+2.  Ask the SecurityContextConstraintsAllocator to create the security context
+3.  Ask the SecurityContextConstraintsAllocator to create the security context.  The allocator
+will not overwrite fields already set in the container.
+4.  Validate that the generated SecurityContext falls within the boundaries of the security context
+constraints and accept or reject the pod.
+
+#### Creation of a Security Context Based on Security Context Constraints
+
+The creation of a security context based on security context constraints is based upon the configured
+settings of the security context constraints and the cluster wide strategies enforce by the
+allocator (discussed below).
+
+There are three scenarios under which a security context constraint field may fall:
+
+1.  Governed by a boolean: fields of this type will be defaulted to the most restrictive value.
+For instance, `AllowPrivilegedContainer` will always be set to false if unspecified.
+1.  Goverened by an allowable set: fields of this type will be checked against the set to ensure
+their value is allowed.  For example, `AllowedCapabilities` will ensure that only capabilities
+that are allowed to be requested are considered valid.  `HostNetworkSources` will ensure that
+only pods created from source X are allowed to request access to the host network.
+1.  Governed by a strategy: Items that have a strategy to generate a value will provide a
+mechanism to generate the value as well as a mechanism to ensure that a specified value falls into
+the set of allowable values.  See the Types section for the description of the interfaces that
+strategies must implement.
+
+Some items of the security context only need to be allocated a single time.  For instance if the
+cluster defines that each namespace should have a block of UIDs used for allocation to service
+accounts then the allocator may allocate UID 1000 to service account X for the lifetime of the
+service account.  This allocation can be stored as an annotation on the namespace level so
+that it is uneditable.  Allocation of a UID to a user may be stored on the user object.  When
```

Comment:
> Allocation of a UID to a user may be stored on the user object.

That's not likely to be true. User objects don't even exist as API objects yet (all we have at this point is the `user.Info` interface)... even if/once they do, the authentication layer is free to authenticate users that are not backed by cluster-scoped API objects (like cert-based users, service accounts using the API, users looked up in LDAP, etc). I think you have to limit yourself to things the `user.Info` interface provides.